cassandra/metadata.py

```diff
@@ -2565,3 +2566,52 @@ def _cql_from_cass_type(cass_type):
         return cass_type.subtypes[0].cql_parameterized_type()
     else:
         return cass_type.cql_parameterized_type()
+
+
+NO_VALID_REPLICA = object()
+
+
+def group_keys_by_host(session, keyspace, table, keys):
+    """
+    Returns a :class:`collections.defaultdict` with the keys grouped per host. This can be
+    used to more accurately group by IN clause or to batch the keys per host.
+
+    If a valid replica is not found for a particular key it will be grouped under
+    :class:`~.NO_VALID_REPLICA`
+
+    Example usage::
+        result = group_keys_by_host(session, "system", "peers",
+                ("peer", "data_center"), (("127.0.0.1", ), ("127.0.0.2", )))
+    """
+    cluster = session.cluster
+
+    partition_keys = cluster.metadata.keyspaces[keyspace].tables[table].partition_key
+
+    serializers = list(types._cqltypes[partition_key.cql_type] for partition_key in partition_keys)
+    keys_per_host = defaultdict(list)
+    for key in keys:
+        serialized_key = list(serializer.serialize(pk, cluster.protocol_version)
+                              for serializer, pk in zip(serializers, key))
+        if len(serialized_key) == 1:
+            routing_key = serialized_key[0]
+        else:
+            routing_key = b"".join(Statement._key_parts_packed(serialized_key))
+        all_replicas = cluster.metadata.get_replicas(keyspace, routing_key)
+        # First check if there are local replicas
+        valid_replicas = list(host for host in all_replicas if
+                              host.is_up and cluster._default_load_balancing_policy.distance(
+                                  host) == HostDistance.LOCAL)
+        if not valid_replicas:
+            valid_replicas = list(host for host in all_replicas if host.is_up)
+            if not valid_replicas:
+                # We will group under this statement all the keys for which
+                # we haven't found a valid replica
+                keys_per_host[NO_VALID_REPLICA].append(key)
+        else:
+            for replica in valid_replicas:
+                if replica in keys_per_host:
+                    keys_per_host[replica].append(key)
+                    break
+            else:
+                keys_per_host[valid_replicas.pop()].append(key)
```

Comment:
I think it could be interesting to not always select blindly the last replica. It would be nice if we could have this done in a random way... but based on the previously selected replica. We can discuss that by voice so I can clarify what I'm thinking of.